Draft: Get started guide xk6 browser (#828)
grafana/k6-docs: src/data/markdown/docs/30 xk6-browser/01 Getting started/03 Running xk6-browser.md

<Blockquote mod="note" title="">

You can also use `page.$()` instead of `page.locator()`. You can find the differences between `page.locator()` and `page.$` in the [Locator API documentation](https://k6.io/docs/javascript-api/xk6-browser/locator/).

</Blockquote>

<CodeGroup labels={["script.js"]} lineNumbers={[true]}>

```javascript
import { chromium } from 'k6/x/browser';

export default function () {
  const browser = chromium.launch({ headless: false });
  const page = browser.newPage();

  page
    .goto('https://test.k6.io/my_messages.php')
    .then(() => {
      // Enter login credentials
      page.locator('input[name="login"]').type('admin');
      page.locator('input[name="password"]').type('123');

      page.screenshot({ path: 'screenshot.png' });
    })
    .finally(() => {
      page.close();
      browser.close();
    });
}
```

</CodeGroup>

The preceding code creates and returns a Locator object with the selectors for both login and password passed as arguments. 

Within the Locator API, various methods such as `type()` can be used to interact with the elements. The `type()` method types a text to an input field. 

## Asynchronous operations

> As explained previously, k6 operations are synchronous. However, Playwright operations are asynchronous in nature and since xk6-browser aims to provide rough compatibility with Playwright, some of our operations are now asynchronous.

Reviewer: It might be good to mention that we're not just transitioning to async to follow Playwright's lead, but because it's a better fit for interacting with the browser.

Also, let's stick to a single way of referring to functions, or methods in this case, instead of "operations", which is a bit ambiguous in this context.

```suggestion
As explained previously, the k6 API is synchronous. However, since many browser operations happen asynchronously, and in order to follow the Playwright API more closely, we are working on migrating most xk6-browser methods to be asynchronous as well.
```

> At the moment, operations such as `page.goto()`, `page.waitForNavigation()` and `Element.click()` return promises and work similarly to [JavaScript promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises).

Reviewer: ```suggestion
At the moment, methods such as `page.goto()`, `page.waitForNavigation()` and `Element.click()` return promises and work similarly to [JavaScript promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises).
```

---
title: 'Running xk6-browser'
excerpt: 'Follow along to learn how to run xk6-browser test, interact with elements on the page, wait for page navigation, write assertions and run both browser-level and protocol-level in a single script.'
---

Follow along to learn how to:

1. Run a test
2. Interact with elements on your webpage
3. Wait for page navigation
4. Run both browser-level and protocol-level test in a single script

<Blockquote mod="note" title="">

With these example snippets, you'll run the test locally with your machine's resources. xk6-browser is not available within k6 cloud as of yet.

</Blockquote>

## Run a test

To run a simple local script:

1. Follow the [Installation Guidelines](/javascript-api/xk6-browser/get-started/installation/) and make sure there is a binary named `xk6-browser` in your current working directory.

2. Copy the following code, paste it into your favorite editor, and save it as `script.js`:

  <CodeGroup labels={["script.js"]} lineNumbers={[true]}>

  ```javascript
  import { chromium } from 'k6/x/browser';

  export default function () {
    const browser = chromium.launch({ headless: false });
    const page = browser.newPage();

    page

>       .goto('https://test.k6.io/')

Reviewer 2: ```suggestion
      .goto('https://test.k6.io/', { waitUntil: 'networkidle' })
```

  To provide rough compatibility with the Playwright API, the xk6-browser API is also being converted from synchronous to asynchronous. `page.goto()` is now asynchronous so `.then()` is used to deal with the asynchronous nature of the operation.

  </Blockquote>

3. Then, run xk6-browser on your terminal with this command:

  <CodeGroup labels={["CLI"]}>

  ```bash
  $ ./xk6-browser run script.js
  ```

  </CodeGroup>

  <Blockquote mod="note" title="">

  The `./` prefix tells your shell to run the binary located in the current working directory. This is required on macOS and Linux, but not on the Windows `cmd.exe` shell. On PowerShell, specify `.\xk6-browser` instead.

  If you installed xk6-browser with a system package, or placed the binary in a directory that's part of your `$PATH` environment variable, you can omit the `./` or `.\` prefixes.

  </Blockquote>

## Interact with elements on your webpage

You can use `page.locator()` and pass in the element's selector you want to find on the page. `page.locator()` will create and return a [Locator](/javascript-api/xk6-browser/locator/) object, which you can later use to interact with the element.

> To find out which selectors xk6-browser supports, check out [Selecting Elements](/javascript-api/xk6-browser/get-started/selecting-elements/).

Author: Ah yes, will fix!
